ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.logging.processing.NoopProcessingLogContext;
+import io.confluent.ksql.metastore.model.StructuredDataSource;
+import io.confluent.ksql.parser.tree.AstVisitor;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.InsertValues;
+import io.confluent.ksql.parser.tree.Literal;
+import io.confluent.ksql.parser.tree.Node;
+import io.confluent.ksql.parser.tree.NullLiteral;
+import io.confluent.ksql.rest.entity.KsqlEntity;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.SchemaUtil;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+
+public class InsertValuesExecutor {
+
+  private static final long MAX_SEND_TIMEOUT_SECONDS = 5;
+
+  private final LongSupplier clock;
+
+  public InsertValuesExecutor() {
+    this(System::currentTimeMillis);
+  }
+
+  @VisibleForTesting
+  InsertValuesExecutor(final LongSupplier clock) {
+    this.clock = Objects.requireNonNull(clock, "clock");
+  }
+
+  public Optional<KsqlEntity> execute(
+      final ConfiguredStatement<InsertValues> statement,
+      final KsqlExecutionContext executionContext,
+      final ServiceContext serviceContext
+  ) {
+    final InsertValues insertValues = statement.getStatement();
+    final StructuredDataSource<?> dataSource = executionContext
+        .getMetaStore()
+        .getSource(insertValues.getTarget().getSuffix());
+
+    if (dataSource.hasWindowedKey()) {
+      throw new KsqlException("Cannot insert values into windowed stream/table!");
+    }
+
+    final KsqlConfig config = statement.getConfig()
+        .cloneWithPropertyOverwrite(statement.getOverrides());
+
+    final GenericRow row = extractRow(insertValues, dataSource);
+    final byte[] key = serializeKey(row.getColumnValue(SchemaUtil.ROWKEY_NAME_INDEX), dataSource);
+    final byte[] value = serializeRow(row, dataSource, config, serviceContext);
+
+    final String topicName = dataSource.getKafkaTopicName();
+
+    // for now, just create a new producer each time
+    final Producer<byte[], byte[]> producer = serviceContext
+        .getKafkaClientSupplier()
+        .getProducer(config.getProducerClientConfigProps());
+
+    producer.send(
```

Comment:
KStreams will retry certain failure types. It has config to configure the number / delay of retries. We should consider adding retries here too and potentially picking up the streams config to control it.

Reply:
There's a producer config `retries`. If the user wants to, they should be able to set this in the producer config :) 
> Setting a value greater than zero will cause the client to resend any record whose send fails with a potentially transient error. Note that this retry is no different than if the client resent the record upon receiving the error. Allowing retries without setting max.in.flight.requests.per.connection to 1 will potentially change the ordering of records because if two batches are sent to a single partition, and the first fails and is retried but the second succeeds, then the records in the second batch may appear first. Note additionall that produce requests will be failed before the number of retries has been exhausted if the timeout configured by delivery.timeout.ms expires first before successful acknowledgement. Users should generally prefer to leave this config unset and instead use delivery.timeout.ms to control retry behavior.

Reply:
Or better yet, they set `delivery.timeout.ms`

Reply:
Fairy-snuff. :D

Reply:
I will have a big-*** documentation PR at the end of this INSERT INTO stuff, I will make sure to include that there